### PR TITLE
Add NoConstantPropagationAnnotation to disable constatnt propagation

### PR DESCRIPTION
**Pull request by sequencer**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [ ] Did you update the FIRRTL spec to include every new feature/behavior?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement
- new feature/API

#### API Impact
add

#### Backend Code Generation Impact
none

#### Desired Merge Strategy
- Rebase: You will rebase the PR onto master and it will be merged with a merge commit.

#### Release Notes
--no-constant-propagation can be used to disable constant propagation.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (1.2.x, 1.3.0, 1.4.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

File: src/main/scala/firrtl/stage/FirrtlAnnotations.scala [hunk 1]
```diff
@@ -290,24 +290,3 @@ case object PrettyNoExprInlining extends NoTargetAnnotation with FirrtlOption wi
...
-          .orElse(throw new OptionsException(s"Unknown primop '$a'. (Did you misspell it?)"))
-          .map(DisableFold(_))
-          .toSeq
-      },
-      helpText = "Disable folding of specific primitive operations",
-      helpValueName = Some("<primop>")
-    )
-  )
-
-}
```
Comment (sequencer) [inline src/main/scala/firrtl/stage/FirrtlAnnotations.scala:313]: I think it breaks binary compatibility, but this class should belong to FIRRTL privately, and won't affect users experience.

File: src/main/scala/firrtl/transforms/ConstantPropagation.scala [hunk 2]
```diff
@@ -124,6 +124,29 @@ class ConstantPropagation extends Transform with DependencyAPIMigration {
...
+          .toMap
+          .get(a)
+          .orElse(throw new OptionsException(s"Unknown primop '$a'. (Did you misspell it?)"))
+          .map(DisableFold(_))
+          .toSeq
+      },
+      helpText = "Disable folding of specific primitive operations",
+      helpValueName = Some("<primop>")
+    )
+  )
```
Comment (sequencer) [inline src/main/scala/firrtl/transforms/ConstantPropagation.scala:148]: I move that `ShellOption` to here, for better consistency.

File: src/main/scala/firrtl/transforms/ConstantPropagation.scala [hunk 3]
```diff
@@ -101,7 +101,7 @@ object ConstantPropagation {
 
 }
 
-class ConstantPropagation extends Transform with DependencyAPIMigration {
+class ConstantPropagation extends Transform with RegisteredTransform with DependencyAPIMigration {
```
Comment (seldridge) [inline src/main/scala/firrtl/transforms/ConstantPropagation.scala:104]: Can you check if this works? Pass a `--help` to `FirrtlStage` and make sure this shows up. I think if this goes the `RegisteredTransform` route then it needs to add to https://github.com/chipsalliance/firrtl/blob/master/src/main/resources/META-INF/services/firrtl.options.RegisteredTransform. The linkage is annoying, but there's no other way to make something discoverable like this unless you (1) are looking for things that mix-in a sealed trait or (2) very slowly walk all the classes.

File: src/main/scala/firrtl/transforms/ConstantPropagation.scala [hunk 4]
```diff
@@ -869,6 +891,12 @@ class ConstantPropagation extends Transform with DependencyAPIMigration {
 
     val disabledOps = state.annotations.collect { case DisableFold(op) => op }.toSet
 
-    state.copy(circuit = run(state.circuit, dontTouchMap, disabledOps))
+    val noConstantPropagation = state.annotations.contains(NoConstantPropagationAnnotation)
+    if (noConstantPropagation) {
```
Comment (seldridge) [inline src/main/scala/firrtl/transforms/ConstantPropagation.scala:895]: Inline this guy:

```suggestion
    if (state.annotations.contains(NoConstantPropagationAnnotation)) {
```

File: src/main/scala/firrtl/transforms/ConstantPropagation.scala [hunk 5]
```diff
@@ -124,6 +124,28 @@ class ConstantPropagation extends Transform with DependencyAPIMigration {
     case _                      => false
   }
 
+  val options = Seq(
+    new ShellOption[Unit](
+      longOption = "no-constant-propagation",
+      toAnnotationSeq = (_: Unit) => Seq(NoConstantPropagationAnnotation),
+      helpText = "Disable constant propagation elimination"
```
Comment (seldridge) [inline src/main/scala/firrtl/transforms/ConstantPropagation.scala:131]:
```suggestion
      helpText = "Disable constant propagation"
```

File: src/main/scala/firrtl/transforms/ConstantPropagation.scala [hunk 6]
```diff
@@ -124,6 +124,28 @@ class ConstantPropagation extends Transform with DependencyAPIMigration {
     case _                      => false
   }
 
+  val options = Seq(
+    new ShellOption[Unit](
+      longOption = "no-constant-propagation",
+      toAnnotationSeq = (_: Unit) => Seq(NoConstantPropagationAnnotation),
```
Comment (seldridge) [inline src/main/scala/firrtl/transforms/ConstantPropagation.scala:130]: Nit: Does this need the explicit type?

```suggestion
      toAnnotationSeq = _ => Seq(NoConstantPropagationAnnotation),
```

File: src/test/scala/firrtlTests/ConstantPropagationTests.scala [hunk 7]
```diff
@@ -908,6 +908,34 @@ class ConstantPropagationIntegrationSpec extends LowTransformSpec {
...
+        |    inst c0 of Child0
+        |    inst c1 of Child1
+        |    z0 <= c0.out
+        |    z1 <= c1.out
+        |    c0.in0 <= UInt<1>(1)
+        |    c0.in1 <= UInt<1>(1)
+        |    c1.in0 <= UInt<1>(0)
+        |    c1.in1 <= UInt<1>(0)""".stripMargin
+    val check = parse(input).serialize
+    execute(input, check, Seq(NoConstantPropagationAnnotation))
```
Comment (seldridge) [inline src/test/scala/firrtlTests/ConstantPropagationTests.scala:936]: This test case is pretty big for what it's checking. Can you trim this down to just one module? Something like the following:

```scala
circuit Foo:
  module Foo:
    input a: UInt<1>
    output b: UInt<1>
    b <= and(UInt<1>, a)
```

File: src/main/scala/firrtl/transforms/ConstantPropagation.scala (hunk 3 above)
Comment (sequencer) [reply src/main/scala/firrtl/transforms/ConstantPropagation.scala:104]: Sorry, I didn't check it actually, I have fixed that in up-to-date commits.

File: src/test/scala/firrtlTests/ConstantPropagationTests.scala (hunk 7 above)
Comment (sequencer) [reply src/test/scala/firrtlTests/ConstantPropagationTests.scala:936]: Sure!

File: src/main/scala/firrtl/stage/FirrtlAnnotations.scala (hunk 1 above)
Comment (seldridge) [reply src/main/scala/firrtl/stage/FirrtlAnnotations.scala:313]: Somehow I seem to have lost my comment about this. 

I'd like to get this into 1.4.3. Annoyingly, this `DisableFold` thing was already in 1.4.2... I agree that this is entirely irrelevant to user experience. However, I'm concerned about breaking binary compatibility here.

What about leaving this in and deprecating it?

**Review by seldridge — COMMENTED** [review]
This is looking great, @sequencer! 

I'm fine to merge this if we can get around the binary compatibility issue. I'd like this to get into 1.4.3 and not have to wait until 1.5.0 for this to show up. I'm also hesitant to waive binary compatibility checks for this.

**Comment by sequencer** [comment]
Thanks @seldridge, I just rebased the binary incompatible part to #2152. I think this PR is safe to merge now.

**Review by jackkoenig — APPROVED** [review]

**Pull request merged**